tests/keras/backend/backend_test.py

```diff
@@ -1040,6 +1041,37 @@ def test_conv(self, op, input_shape, kernel_shape, padding, data_format):
             padding=padding, data_format=data_format,
             cntk_dynamicity=True)
 
+    @pytest.mark.skipif((KC is not None and KC.dev.type() == 0),
+                        reason='cntk only supports dilated conv on GPU')
+    @pytest.mark.parametrize('op,input_shape,kernel_shape,padding,data_format,dilation_rate', [
+        ('conv1d', (2, 8, 3), (4, 3, 2), 'valid', 'channels_last', 2),
+        ('conv1d', (2, 3, 8), (4, 3, 2), 'valid', 'channels_first', 2),
+        ('conv2d', (2, 8, 9, 3), (3, 3, 3, 2), 'same', 'channels_last', (2, 2)),
+        ('conv2d', (2, 3, 9, 8), (4, 3, 3, 4), 'valid', 'channels_first', (2, 2)),
+        ('conv3d', (2, 5, 4, 6, 3), (2, 2, 3, 3, 4), 'valid', 'channels_last', (2, 2, 2)),
+        ('conv3d', (2, 3, 5, 4, 6), (2, 2, 3, 3, 4), 'same', 'channels_first', (2, 2, 2)),
+    ])
+    def test_conv_dilation(self, op, input_shape, kernel_shape, padding,
+                           data_format, dilation_rate):
+        check_two_tensor_operation(
```

Comment:
I think we can't use this line ```backends_to_test = [KTH, KTF]```. We know TF and TH installed on travis, but all users may download Keras code and run the tests, they may only have TF or TH installed, then the other one would be ```None``` and throw exceptions when running the tests. I think this would make users confused. And I think last changed we made at
```
@pytest.mark.skipif((K.backend() == 'cntk' and K.dev.type() == 0),
                    reason='cntk only supports dilated conv on GPU')
```
is want to let the tests can be run if users download the code to a node has CNTK with GPU.

Reply:
If we really want to run this test as much as possible, I would suggest to add
```
        backend_list = []
        if KTF is not None:
            backend_list.append(KTF)
        if KTH is not None:
            backend_list.append(KTH)
        if KC is not None and KC.dev.type() == 1:
            backend_list.append(KC)
```
It looks a little ugly, but it does work. What do you think of it?

Reply:
I think that you're right and this is our best option until the numpy backend `KNP` can handle all types of convolutions.